File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];
rl.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  function find(now) {
    const idx = now[now.length - 1];
    const nextNode = l[idx];
    if (visited[nextNode]) {
      while (now.length) {
        if (now[0] == nextNode) {
          answer += now.length;
          break;
        }
        now.shift();
      }
    } else if (!visited[nextNode]) {
      visited[nextNode] = true;
      now.push(nextNode);
      find(now);
    }
  }

  const T = Number(input.shift());
  let N = 0;
  let l = [];
  let visited = [];
  let answers = [];
  let answer = 0;
  for (let i = 0; i < T; i++) {
    N = Number(input.shift());
    l = [0].concat(input.shift().split(' ').map(Number));
    for (let i = 0; i <= N; i++) {
      visited.push(false);
    }
    for (let i = 1; i <= N; i++) {
      if (!visited[i]) {
        visited[i] = true;
        find([i]);
      }
    }
    answers.push(N - answer);
    l = [];
    answer = 0;
    visited = [];
  }
  console.log(answers.join('\n'));
  process.exit();
});

File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.py
import sys
input=sys.stdin.readline

T=int(input())
for _ in range(T):
  n=int(input())
  l=[0]+list(map(int,input().split()))

  team=0
  visited=[False for i in range(n+1)]

  for i in range(1,n+1):
    if visited[i]:
      continue

    visited[i]=True
    s=[i]
    now=i
    while True:
      if visited[l[now]]:
        cnt=0
        while s:
          item=s.pop()
          cnt+=1
          if item==l[now]:
            team+=cnt
            break
        break
      else:
        visited[l[now]]=True
        s.append(l[now])
        now=l[now]

  print(n-team)
